identity/aziot-identityd/src/lib.rs

```diff
@@ -702,76 +634,204 @@ impl Api {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let private_key = openssl::pkey::PKey::from_rsa(rsa).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let private_key_pem =
+                        private_key.private_key_to_pem_pkcs8().map_err(|err| {
+                            Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                        })?;
+                    let private_key_pem =
+                        std::string::String::from_utf8(private_key_pem).map_err(|err| {
+                            Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                        })?;
+                    let public_key = private_key.public_key_to_pem().map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let public_key = openssl::pkey::PKey::public_key_from_pem(&public_key)
+                        .map_err(|err| {
                             Error::Internal(InternalError::CreateCertificate(Box::new(err)))
                         })?;
 
-                    (identity_public_key, identity_private_key)
-                };
-
-                let csr = create_csr(
-                    &subject,
-                    &identity_public_key,
-                    &identity_private_key,
-                    attributes,
-                )
-                .map_err(|err| Error::Internal(InternalError::CreateCertificate(Box::new(err))))?;
-                csr
+                    (private_key, private_key_pem, public_key)
+                }
             };
 
-            let _new_cert = self
-                .cert_client
-                .create_cert(&identity_cert, &csr, None)
-                .await
-                .map_err(|err| Error::Internal(InternalError::CreateCertificate(Box::new(err))))?;
-        }
+            let result = async {
+                let csr =
+                    create_csr(&subject, &public_key, &private_key, attributes).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+
+                let cert = self
+                    .cert_client
+                    .create_cert(&identity_cert, &csr, None)
+                    .await
+                    .map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                device_id_cert = Some(cert);
+
+                Ok::<(), Error>(())
+            }
+            .await;
 
-        Ok(())
+            if let Err(err) = result {
+                if let KeyType::Handle(_key_id) = identity_pk {
+                    // TODO: need to delete key from keyd.
```

Comment:
Implement the TODO?

Reply:
Not possible right now -- there's no KS API to delete keys.